### DISCUSSION TITLE
Rootless k3s stuck at "Waiting for containerd startup: rpc error: code = Unimplemented desc = unknown service runtime.v1.RuntimeService"

### DISCUSSION THREAD

**Discussion by edysli** [k3s-io/k3s#11304] [Q&A]
I (finally) managed to get a rootless k3s to start, but it continuously prints the same two lines in the log and makes no progress:
```
k3s[27860]: time="2024-11-12T21:21:01+01:00" level=info msg="Waiting for containerd startup: rpc error: code = Unimplemented desc = unknown service runtime.v1.RuntimeService"
k3s[27860]: time="2024-11-12T21:21:01+01:00" level=info msg="Waiting for control-plane node mypc startup: nodes \"mypc\" not found"
```
It seems containerd's socket path is wrongly set, because earlier in the log, there is:
```
k3s[27860]: time="2024-11-12T21:20:10+01:00" level=info msg="Running containerd -c /home/user/.rancher/k3s/agent/etc/containerd/config.toml -a /run/k3s/containerd/containerd.sock --state /run/k3s/containerd --root /home/user/.rancher/k3s/agent/containerd"
```
and there is nothing in `/run/k3s/containerd/`, instead the socket is `/run/user/1000/k3s/containerd/containerd.sock`.

Setting `kubelet-arg: - "container-runtime-endpoint=unix:///run/user/1000/k3s/containerd/containerd.sock"` in my `config.yaml` doesn't fix the issue. k3s still uses a hardcoded path `/run/k3s/containerd/containerd.sock` which doesn't exist.

Any idea how to fix this?

**Comment by brandond** [comment]
The problem is not that the socket path is wrong. If it was the wrong path you wouldn't be able to connect to it to get that "unknown service" RPC error from the server. Note that many of the paths reported by rootless K3s will appear to not exist when you're poking at it from outside. These paths only exist within the k3s service's filesystem namespace.

The problem is that containerd is running, but the CRI service is not enabled for some reason. Check containerd.log to see why. Usually it has something to do with a bad snapshotter causing the service setup to fail.

Why are you having so much trouble with rootless K3s? Running it as a systemd user service, as described in the docs, should work fine.

**Reply by edysli** [reply]
Thank you for the hint. There is indeed a CRI-related failure in `containerd.log`, but it's disguised as a warning, thus I missed it yesterday:
```
time="2024-11-12T21:20:11.060559192+01:00" level=warning msg="failed to load plugin io.containerd.grpc.v1.cri" error="failed to create CRI service: failed to create cni conf monitor for default: failed to create cni conf dir=/etc/cni/net.d for watch: mkdir /etc/cni/net.d: permission denied"
```
Is containerd trying to read the real `/etc/cni/net.d/` directory? It exists on my system, but is only accessible to root (0700).

I'm having much trouble with rootless k3s (or rootful too, for that matter), because I want to run a single-stack IPv6 cluster (no IPv4 at all) and the Docker & Kubernetes world is notoriously bad at IPv6. For example, you can't use a /64 network because some component cannot cope with IP address allocation on anything bigger than a /112 (kubelet or cilium or something else... it's neither clear nor documented).
Running rootless made it slightly worse because slirp4netns's IPv6 support is experimental and apparently cannot work without IPv4, it must have a configuration for it. `slirp4netns --cidr` only supports IPv4 ranges and this affects `K3S_ROOTLESS_CIDR`. How do I configure the IPv6 network I want to use? This is apparently impossible. I've added an IPv4 network in the configuration just to get k3s to start, and then had to endlessly fumble around with IPv6 and IPv4 addresses and their order in `cluster-cidr`, `service-cidr`, and `node-ip` until no more errors appeared in k3s's log. By the way, it's impossible to tell from k3s's log which built-in Kubernetes component throws an error because the component's name doesn't appear on the log line.

**Reply by edysli** [reply]
Oh and that bug https://github.com/k3s-io/k3s/issues/2420#issuecomment-715051120 bites me every time I start k3s-rootless.service :/ Why would I enable IPv4 forwarding when I only want IPv6?

**Reply by brandond** [reply]
> For example, you can't use a /64 network because some component cannot cope with IP address allocation on anything bigger than a /112 (kubelet or cilium or something else... it's neither clear nor documented).

It's the controller-manager. A /64 is WAY too big for the cluster-cidr, since the IPAM uses a bitmask to store used addresses, it would require a stupid amount of memory to store half of the total IPv6 address space. Why are you trying to use such a large range?

> Running rootless made it slightly worse because slirp4netns's IPv6 support is experimental and apparently cannot work without IPv4, it must have a configuration for it. slirp4netns --cidr only supports IPv4 ranges and this affects K3S_ROOTLESS_CIDR.

Yep.

> it's impossible to tell from k3s's log which built-in Kubernetes component throws an error because the component's name doesn't appear on the log line.

This is mostly due to Kubernetes sharing a LOT of code. Without passing a logging context around (which Kubernetes does not do) there is no way to tell which component is calling into the shared code that actually logs the message. If this is a problem, you might consider a different distro that runs the components separately.

> Why would I enable IPv4 forwarding when I only want IPv6?

We fixed this ages go. It'll now only enable forwarding for address families used by the service. What version of K3s are you even using?

**Reply by brandond** [reply]
You should also investigate how slirp4netns actually works. The IPv4 network you're worrying about here is JUST used for the two ends of the tun/tap interfaces that pass traffic between the host and user namespaces. If you want it to support pure ipv6 operation, you should take that to the upstream projects:
https://github.com/rootless-containers/slirp4netns
https://github.com/rootless-containers/rootlesskit/tree/master/pkg/network/slirp4netns

**Reply by edysli** [reply]
> Why are you trying to use such a large range?

Because /64 is a pretty common network size in the v6 world and actually the smallest one can get, because the lower 64 bits are supposed to be the interface identifier. I had no idea prefixes larger than 64 could even work. :open_mouth: 

> We fixed this ages go. It'll now only enable forwarding for address families used by the service. What version of K3s are you even using?

Ah that one probably appeared when I added IPv4 addresses to cluster-cidr, service-cidr, and node-ip to get slirp4netns to work... not sure exactly when it appeared though.
```
$ k3s --version
k3s version v1.30.5+k3s1 (9b586704)
go version go1.22.6
```

Coming back to containerd, does the rootless version really need access to the host's `/etc/cni/net.d/` directory?

**Reply by brandond** [reply]
I've never seen that particular error. Have you tried just deleting it from the host so that it can be created within the user filesystem namespace?

**Reply by edysli** [reply]
So I went ahead and removed `/etc/cni/` and I'm hitting a different error (progress! :smile: ). containerd now clearly logs a failure:
```
time="2024-11-13T22:11:20.363968806+01:00" level=error msg="failed to load cni during init, please check CRI plugin status before setting up network for pods" error="cni config load failed: no network config found in /etc/cni/net.d: cni plugin not initialized: failed to load cni config"
```
and k3s logs a different error now too (repeatedly)
```
k3s[11665]: E1113 22:33:58.268538      56 kubelet_node_status.go:711] "Failed to set some node status fields" err="failed to validate nodeIP: node IP: \"fd5b:e9ce:dffb:7ea::100\" not found in the host's network interfaces" node="mypc"
```
I conclude containerd expects a CNI configuration at `/etc/cni/`. This differs from the k3s "rootful" installation where this configuration was automatically created (by Cilium, in my case). I does indeed work if I make `/etc/cni/` readable (except Cilium isn't there yet).

**Reply by brandond** [reply]
You're running in a network namespace. The rootless eth0 IP is not the host's IP. It is the IP of the rootless side of the slirp4netns tunnel device, the one you were complaining about being ipv4-only.

If you're using another CNI besides the flannel that comes with k3s, it won't be aware of k3s's CNI directory. It'll use whatever its defaults are.

I **really** don't think you're going to get cilium working rootless.

**Reply by edysli** [reply]
## IPv6 first
So I removed Cilium from the picture and enabled the built-in Flannel to simply things. However, I still get IPv6-related failures with this configuration:
```yaml
cluster-cidr:
  - "fd5b:e9ce:dffb:cafe::/64"
  - "192.168.33.0/24"
service-cidr:
  - "fd5b:e9ce:dffb:07ea::/112"
  - "192.168.32.0/24"
node-ip:
  - "fd5b:e9ce:dffb:07ea::100"
  - "192.168.31.100"
```
and `K3S_ROOTLESS_ENABLE_IPV6=true K3S_ROOTLESS_CIDR=192.168.31.0/24`.
k3s repeatedly logs these three lines:
```
time="2024-11-14T22:51:54+01:00" level=info msg="Connecting to proxy" url="wss://[fd5b:e9ce:dffb:7ea::100]:6443/v1-k3s/connect"
time="2024-11-14T22:51:54+01:00" level=error msg="Failed to connect to proxy. Empty dialer response" error="dial tcp [fd5b:e9ce:dffb:7ea::100]:6443: connect: network is unreachable"
time="2024-11-14T22:51:54+01:00" level=error msg="Remotedialer proxy error; reconnecting..." error="dial tcp [fd5b:e9ce:dffb:7ea::100]:6443: connect: network is unreachable" url="wss://[fd5b:e9ce:dffb:7ea::100]:6443/v1-k3s/connect"
```

## IPv4 first
Weird thing is, if I put the IPv4 addresses first in the configuration:
```yaml
cluster-cidr:
  - "192.168.33.0/24"
  - "fd5b:e9ce:dffb:cafe::/64"
service-cidr:
  - "192.168.32.0/24"
  - "fd5b:e9ce:dffb:07ea::/112"
node-ip:
  - "192.168.31.100"
  - "fd5b:e9ce:dffb:07ea::100
```
The errors above don't occur and it progresses to another error... :confused:
This time it's not network related:
```
E1114 23:07:09.597096      53 remote_runtime.go:343] "StartContainer from runtime service failed" err="rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting \"/home/edysli/.rancher/k3s/agent/kubelet/pods/45a40c1d-d11c-42eb-9912-e2b008d28a74/volumes/kubernetes.io~configmap/content\" to rootfs at \"/chart\": mount /home/edysli/.rancher/k3s/agent/kubelet/pods/45a40c1d-d11c-42eb-9912-e2b008d28a74/volumes/kubernetes.io~configmap/content:/chart (via /proc/self/fd/6), flags: 0x5021: operation not permitted: unknown" containerID="587d45fa5b3f055ca7ec40f739027f4a7e352443a20e85e812c282e0d317c978"
```
All pods are stuck in CrashLoopBackOff.
I do have `/home` mounted with nosuid and nodev. Could this be related to https://github.com/k3s-io/k3s/discussions/10851 ?

**Reply by brandond** [reply]
>`K3S_ROOTLESS_CIDR=192.168.31.0/24`
>```yaml
>service-cidr:
>  - "fd5b:e9ce:dffb:07ea::/112"
>  - "192.168.32.0/24"
>node-ip:
>  - "fd5b:e9ce:dffb:07ea::100"
>  - "192.168.31.100"
>```

What are you doing here? Why are you trying to use an address from the cluster service cidr as the node's ipv6 address? How would that work?

It looks like slirp4nets is [hardcoded](https://github.com/rootless-containers/slirp4netns/blob/ee1542e1532e6a7f266b8b6118973ab3b10a8bb5/slirp4netns.c#L263) to use `fd00::1` as the interface's ipv6 address, it cannot be changed like the IPv4 CIDR. I'm kinda guessing here as I am not super familiar with libslirp's [SlirpConfig](https://gitlab.freedesktop.org/slirp/libslirp/-/blob/master/src/libslirp.h?ref_type=heads#L179-186).



The node-ip options tell Kubernetes what addresses to use; those IPs have to actually be present on an interface, you can't just pick random IPs.

> I do have /home mounted with nosuid and nodev. 

Yeah that's probably not going to work either. Maybe try using something outside of /home as your data-dir, if you don't want to change that? I'm also assuming /home isn't on a NFS share or something else even less likely to work?

**Reply by edysli** [reply]
> What are you doing here? Why are you trying to use an address from the cluster service cidr as the node's ipv6 address? How would that work?

because I am completely clueless about configuring `node-ip` and `node-external-ip`... :shrug:
[`k3s agent`'s documentation](https://docs.k3s.io/cli/agent#networking) says:
> `--node-ip` IP address to advertise for node
> `--node-external-ip` External IP address to advertise for node

[kubelet's documentation](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) explains a bit more:
> IP address (or comma-separated dual-stack IP addresses) of the node. If unset, kubelet will use the node's default IPv4 address, if any, or its default IPv6 address if it has no IPv4 addresses. You can pass "::" to make it prefer the default IPv6 address rather than the default IPv4 address.

Neither clearly explains what that address is used for nor how to pick one. Of course, I've tried leaving this unconfigured and tried `::` too, none of these left me with a working network (same proxy errors as above). Without configuration, 192.168.31.100 gets picked up magically, now I suppose that is one end of the slirp4netns connection. It's hard to understand what's going on when I can't see any of the network interfaces in that namespace... Sorry for fumbling in the dark and bumping into the furniture.

> I'm also assuming /home isn't on a NFS share or something else even less likely to work?

No, it's not NFS just ext4.
Are there issues with ZFS? I've seen containerd must be configured with a special snapshotter for ZFS.

**Reply by brandond** [reply]
Normally containerd just uses overlayfs which should work on most filesystems. If the filesystem itself supports snapshotting (like zfs does) then you can use a filesystem-specific snapshotter. If you can't use anything else (not even overlayfs) is the native snapshotter but it is super inefficient because it just unpacks all the layers on top of each other without any potential for reuse.

You're jumping in to the deep end by trying to get all this working without understanding Kubernetes and containerd and Linux namespaces work at a low level. Rootless Kubernetes works within very tightly defined constraints and there is a lot of stuff that's not possible without someone writing a bunch of additional code to wire up all the different parts.
